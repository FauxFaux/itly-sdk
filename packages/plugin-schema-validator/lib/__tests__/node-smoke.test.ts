/**
 * Test for Node version of plugin-schema-validator
 * @jest-environment node
 */
/* eslint-disable no-unused-vars, no-console */
/* eslint-disable import/no-unresolved, import/extensions, import/no-dynamic-require */
import Itly, {
  Options, Plugin, Event, ValidationResponse,
} from '@itly/sdk';
import CustomPlugin from '../../../../__tests__/src/CustomPlugin';
import {
  TestParams,
  requireForTestEnv,
} from '../../../../__tests__/util';

const SchemaValidator = requireForTestEnv(__dirname);

const testSchemas = require('../../../../__tests__/data/basic-schema.json');

const context = {
  requiredString: 'is required',
  optionalEnum: 'Value 1',
};

const identifyProps = {
  requiredNumber: 42,
};

const plugins: Plugin[] = [
  new SchemaValidator(testSchemas),
  new CustomPlugin(),
];

const testParams: TestParams[] = [
  {
    name: 'load, track, validate - validationOptions=DEFAULT',
    context,
    options: {
      environment: 'production',
      plugins,
    },
  },
  {
    name: 'load, track, validate - validationOptions={trackInvalid: true}',
    context,
    options: {
      environment: 'production',
      plugins,
      validation: {
        disabled: false,
        trackInvalid: true,
        errorOnInvalid: false,
      },
    },
  },
  {
    name: 'load, track, validate - validationOptions={errorOnInvalid: true}',
    context,
    options: {
      environment: 'production',
      plugins,
      validation: {
        disabled: false,
        trackInvalid: false,
        errorOnInvalid: true,
      },
    },
  },
  {
    name: 'load, track, validate w/ context=\'undefined\'',
    context: undefined,
    options: {
      environment: 'production',
      plugins,
    },
  },
  {
    name: 'load, track, validate w/ context=\'{}\'',
    context: {},
    options: {
      environment: 'production',
      plugins,
    },
  },
];

// Test vars
const userId = 'test-user-id';
const tempUserId = 'temp-user-id';
const groupId = 'test-group-id';
let spyConsoleLog: jest.SpyInstance;

beforeEach(() => {
  jest.resetModules();
  spyConsoleLog = jest.spyOn(console, 'log').mockImplementation();
});

afterEach(() => {
  spyConsoleLog.mockRestore();
});

test.each(testParams.map((test) => [test.name, test]) as any[])('%s',
  async (name: string, { context: ctx, options }: TestParams) => {
    const itly = new Itly();

    itly.load({
      ...options,
      context: ctx,
    });

    itly.identify(undefined, identifyProps);
    itly.identify(tempUserId, identifyProps);
    try {
      itly.identify(tempUserId);
    } catch (e) {
      console.log('Caught validation error.', e.message);
    }
    try {
      itly.identify(tempUserId, {
        badProp: 'unsupported property',
      });
    } catch (e) {
      console.log('Caught validation error.', e.message);
    }

    itly.alias(userId);
    itly.alias(userId, tempUserId);

    try {
      itly.group(userId, groupId);
    } catch (e) {
      console.log('Caught validation error.', e.message);
    }
    itly.group(userId, groupId, {
      requiredBoolean: true,
      optionalString: undefined,
    });

    try {
      itly.page(userId, 'page category 1', 'page name 1');
    } catch (e) {
      console.log('Caught validation error.', e.message);
    }

    try {
      itly.page(userId, 'page category 2', 'page name 2', {
        pageProp: 'a page property',
      });
    } catch (e) {
      console.log('Caught validation error.', e.message);
    }

    itly.track(userId, {
      name: 'Event No Properties',
      properties: {},
    });

    itly.track(userId, {
      name: 'Event With All Properties',
      properties: {
        requiredString: 'A required string',
        requiredNumber: 2.0,
        requiredInteger: 42,
        requiredEnum: 'Enum1',
        requiredBoolean: false,
        requiredConst: 'some-const-value',
        requiredArray: ['required', 'array'],
        optionalString: 'I\'m optional!',
      },
    });

    try {
      itly.track(userId, {
        name: 'EventMaxIntForTest',
        properties: {
          intMax10: 20,
        },
      });
    } catch (e) {
      console.log('Caught validation error.', e.message);
    }

    expect(spyConsoleLog.mock.calls).toMatchSnapshot();
  });
