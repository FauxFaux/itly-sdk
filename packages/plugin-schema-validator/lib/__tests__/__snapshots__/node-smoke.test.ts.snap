// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load, track, validate - validationOptions={errorOnInvalid: true} 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "load() environment='production' disabled=undefined plugins=[schema-validator, custom] context={\\"requiredString\\":\\"is required\\",\\"optionalEnum\\":\\"Value 1\\"} validation={\\"disabled\\":false,\\"trackInvalid\\":false,\\"errorOnInvalid\\":true}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='undefined' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "Caught validation error.",
    "Validation Error: Passed in identify properties did not validate against your tracking plan. \`properties\` should have required property 'requiredNumber'.",
  ],
  Array [
    "Caught validation error.",
    "Validation Error: Passed in identify properties did not validate against your tracking plan. \`properties\` should NOT have additional properties (badProp).",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='undefined'",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='temp-user-id'",
  ],
  Array [
    "Caught validation error.",
    "Validation Error: Passed in group properties did not validate against your tracking plan. \`properties\` should have required property 'requiredBoolean'.",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties={\\"requiredBoolean\\":true}",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 1' name='page name 1' properties=undefined",
  ],
  Array [
    "Caught validation error.",
    "Validation Error: 'page' schema is empty but properties were found. properties={\\"pageProp\\":\\"a page property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event No Properties' properties={\\"requiredString\\":\\"is required\\",\\"optionalEnum\\":\\"Value 1\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "Caught validation error.",
    "Validation Error: Passed in EventMaxIntForTest properties did not validate against your tracking plan. \`properties.intMax10\` should be <= 10.",
  ],
]
`;

exports[`load, track, validate - validationOptions={trackInvalid: true} 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "load() environment='production' disabled=undefined plugins=[schema-validator, custom] context={\\"requiredString\\":\\"is required\\",\\"optionalEnum\\":\\"Value 1\\"} validation={\\"disabled\\":false,\\"trackInvalid\\":true,\\"errorOnInvalid\\":false}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='undefined' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"badProp\\":\\"unsupported property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='undefined'",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='temp-user-id'",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties={\\"requiredBoolean\\":true}",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 1' name='page name 1' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 2' name='page name 2' properties={\\"pageProp\\":\\"a page property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event No Properties' properties={\\"requiredString\\":\\"is required\\",\\"optionalEnum\\":\\"Value 1\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='EventMaxIntForTest' properties={\\"requiredString\\":\\"is required\\",\\"optionalEnum\\":\\"Value 1\\",\\"intMax10\\":20}",
  ],
]
`;

exports[`load, track, validate - validationOptions=DEFAULT 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "load() environment='production' disabled=undefined plugins=[schema-validator, custom] context={\\"requiredString\\":\\"is required\\",\\"optionalEnum\\":\\"Value 1\\"} validation=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='undefined' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"badProp\\":\\"unsupported property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='undefined'",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='temp-user-id'",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties={\\"requiredBoolean\\":true}",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 1' name='page name 1' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 2' name='page name 2' properties={\\"pageProp\\":\\"a page property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event No Properties' properties={\\"requiredString\\":\\"is required\\",\\"optionalEnum\\":\\"Value 1\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='EventMaxIntForTest' properties={\\"requiredString\\":\\"is required\\",\\"optionalEnum\\":\\"Value 1\\",\\"intMax10\\":20}",
  ],
]
`;

exports[`load, track, validate w/ context='{}' 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "load() environment='production' disabled=undefined plugins=[schema-validator, custom] context={} validation=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='undefined' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"badProp\\":\\"unsupported property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='undefined'",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='temp-user-id'",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties={\\"requiredBoolean\\":true}",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 1' name='page name 1' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 2' name='page name 2' properties={\\"pageProp\\":\\"a page property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event No Properties' properties={}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='EventMaxIntForTest' properties={\\"intMax10\\":20}",
  ],
]
`;

exports[`load, track, validate w/ context='undefined' 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "load() environment='production' disabled=undefined plugins=[schema-validator, custom] context=undefined validation=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='undefined' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"requiredNumber\\":42}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='temp-user-id' properties={\\"badProp\\":\\"unsupported property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='undefined'",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='temp-user-id'",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='test-user-id' groupId='test-group-id' properties={\\"requiredBoolean\\":true}",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 1' name='page name 1' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='test-user-id' category='page category 2' name='page name 2' properties={\\"pageProp\\":\\"a page property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event No Properties' properties={}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='EventMaxIntForTest' properties={\\"intMax10\\":20}",
  ],
]
`;
